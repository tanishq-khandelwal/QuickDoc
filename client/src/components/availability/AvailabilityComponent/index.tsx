import React from "react";
import { AvailabilityComponentProps } from "../types";

const AvailabilityComponent: React.FC<AvailabilityComponentProps> = ({
  weekDays,
  availability,
  timeSlots,
  toggleDropdown,
  handleTimeChange,
  errors,
  setAvailability,
}) => {
  return (
    <div className="space-y-4">
      {weekDays.map((day) => (
        <div key={day.id} className="flex items-center justify-between">
          <div className="flex items-center gap-4">
            <input
              type="checkbox"
              checked={availability[day.id]?.selected || false}
              onChange={() =>
                setAvailability((prev) => ({
                  ...prev,
                  [day.id]: {
                    ...prev[day.id],
                    selected: !prev[day.id].selected,
                  },
                }))
              }
              className="w-5 h-5 accent-blue-500 cursor-pointer"
            />
            <span className="text-lg font-medium">{day.title}</span>
          </div>
          {!availability[day.id]?.selected ? (
            <div className="text-gray-600">Unavailable</div>
          ) : (
            <div className="flex items-center gap-2">
              {/* Start Time Input */}
              <div className="relative">
                <input
                  type="text"
                  value={availability[day.id]?.startTime || "09:00 AM"}
                  onClick={() => toggleDropdown(day.id, "startTime")}
                  onChange={(e) =>
                    handleTimeChange(day.id, "startTime", e.target.value)
                  }
                  className="w-28 h-10 border rounded-lg px-2 text-sm cursor-pointer"
                />
                {availability[day.id]?.openDropdown === "startTime" && (
                  <div className="absolute mt-1 w-28 bg-white border border-gray-300 shadow-lg rounded-md max-h-40 overflow-y-auto z-10">
                    <ul className="space-y-1">
                      {timeSlots?.map((time, index) => (
                        <li
                          key={index}
                          className="dropdown-item px-2 py-1 text-sm cursor-pointer hover:bg-gray-100"
                          onClick={() => {
                            handleTimeChange(day.id, "startTime", time);
                            toggleDropdown(day.id, null);
                          }}
                        >
                          {time}
                        </li>
                      ))}
                    </ul>
                  </div>
                )}
              </div>

              <span className="text-lg font-medium">-</span>

              {/* End Time Input */}
              <div className="relative">
                <input
                  type="text"
                  value={availability[day.id]?.endTime || "09:00 PM"}
                  onClick={() => toggleDropdown(day.id, "endTime")}
                  onChange={(e) =>
                    handleTimeChange(day.id, "endTime", e.target.value)
                  }
                  className="w-28 h-10 border rounded-lg px-2 text-sm cursor-pointer"
                />
                {availability[day.id]?.openDropdown === "endTime" && (
                  <div className="absolute mt-1 w-28 bg-white border border-gray-300 shadow-lg rounded-md max-h-40 overflow-y-auto z-10">
                    <ul className="space-y-1">
                      {timeSlots.map((time, index) => (
                        <li
                          key={index}
                          className="dropdown-item px-2 py-1 text-sm cursor-pointer hover:bg-gray-100"
                          onClick={() => {
                            handleTimeChange(day.id, "endTime", time);
                            toggleDropdown(day.id, null);
                          }}
                        >
                          {time}
                        </li>
                      ))}
                    </ul>
                  </div>
                )}
              </div>
            </div>
          )}
          {/* Display error message if any */}
          {errors[day.id] && (
            <div className="text-red-500 text-sm mt-1">{errors[day.id]}</div>
          )}
        </div>
      ))}
    </div>
  );
};

export default AvailabilityComponent;
