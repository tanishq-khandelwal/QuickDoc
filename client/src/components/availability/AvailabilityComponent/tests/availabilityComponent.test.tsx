import { render, screen } from "@testing-library/react";
import AvailabilityComponent from "..";

describe("AvailabilityComponent", () => {
  const defaultProps = {
    weekDays: [
      { id: 1, title: "Monday" },
      { id: 2, title: "Tuesday" },
    ],
    availability: {
      monday: { selected: false, startTime: "09:00 AM", endTime: "05:00 PM" },
      tuesday: { selected: false, startTime: "09:00 AM", endTime: "05:00 PM" },
    },
    timeSlots: ["09:00 AM", "10:00 AM", "11:00 AM"],
    toggleDropdown: jest.fn(),
    handleTimeChange: jest.fn(),
    errors: {},
    setAvailability: jest.fn(),
  };

  it("Expect to not log errors in console", () => {
    const spy = jest.spyOn(global.console, "error");
    render(<AvailabilityComponent {...defaultProps} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it("renders AvailabilityComponent with default props", () => {
    render(<AvailabilityComponent {...defaultProps} />);
    expect(screen.getByText("Monday")).toBeInTheDocument();
    expect(screen.getByText("Tuesday")).toBeInTheDocument();
    expect(screen.getAllByText("Unavailable").length).toBe(2);
  });

  it("renders with no errors", () => {
    render(<AvailabilityComponent {...defaultProps} />);
    expect(screen.queryByText("Invalid time range")).not.toBeInTheDocument();
  });

  it("renders with all days unavailable", () => {
    render(<AvailabilityComponent {...defaultProps} />);
    expect(screen.getAllByText("Unavailable").length).toBe(2);
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<AvailabilityComponent {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
