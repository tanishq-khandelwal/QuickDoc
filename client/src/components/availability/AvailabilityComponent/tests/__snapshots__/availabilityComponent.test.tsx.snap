// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvailabilityComponent matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="space-y-4"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center gap-4"
      >
        <input
          class="w-5 h-5 accent-blue-500 cursor-pointer"
          type="checkbox"
        />
        <span
          class="text-lg font-medium"
        >
          Monday
        </span>
      </div>
      <div
        class="text-gray-600"
      >
        Unavailable
      </div>
    </div>
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center gap-4"
      >
        <input
          class="w-5 h-5 accent-blue-500 cursor-pointer"
          type="checkbox"
        />
        <span
          class="text-lg font-medium"
        >
          Tuesday
        </span>
      </div>
      <div
        class="text-gray-600"
      >
        Unavailable
      </div>
    </div>
  </div>
</DocumentFragment>
`;
